### PR TITLE
Update membership offline receipt to run off WorkflowTemplate & tokens

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Update membership offline receipt to run off WorkflowTemplate & tokens

Before
----------------------------------------
The offline membership receipt is heavily dependent on form assigns that we are trying to deprecate (the dreaded `dataArray` and the overly-deep `lineItems` as well as an assign for every membership & contribution value.)

It was never very pretty  - no change

image

After
----------------------------------------
Tokens used where possible, otherwise standardised variables.

In addition code is added to permif previewing in the Message Admin screen

Technical Details
----------------------------------------
There is no intended formatting change from this - it is part of the attempt to standardise our templates & decouple them from form logic

Comments
----------------------------------------
@demeritcowboy I am trying to sort out that online receipt - but figured maybe getting the offline one sorted first would make it a bit easier (since at least then I add in the examples) - are you able to take a look once this is passing?  I'd ideally like to make changes to the membership receipts in the same release (5.61) as the membership expiration labels change

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
CRM_Member_Form_MembershipRenewalTest::testSubmitRecur
Exception: CRM_Core_Exception: "Invalid token filter: ["raw"]"
#0 /home/jenkins/bknix-dfl/build/build-0/web/sites/all/modules/civicrm/Civi/Token/TokenProcessor.php(383): Civi\Token\TokenProcessor->filterTokenValue("$ 0.00", (Array:1), Object(Civi\Token\TokenRow), "text/plain")
#1 /home/jenkins/bknix-dfl/build/build-0/web/sites/all/modules/civicrm/Civi/Token/TokenProcessor.php(447): Civi\Token\TokenProcessor->Civi\Token\{closure}("{contribution.total_amount|raw}", "contribution", "total_amount", (Array:1))
#2 [internal function](): Civi\Token\TokenProcessor->Civi\Token\{closure}((Array:4))
#3 /home/jenkins/bknix-dfl/build/build-0/web/sites/all/modules/civicrm/Civi/Token/TokenProcessor.php(431): preg_replace_callback(";\{([\w]+)\.([\w:\.]+)(?:\|(\w+(?::[\w\": %\-_()\[\]\+/#@!,\.\?]*)?))?\};", Object(Closure), "***********************************************************\n\n{ts}Test-drive...")

**Comment by agileware-justin** [comment]
Good one @eileenmcnaughton :1st_place_medal:

**Comment by eileenmcnaughton** [comment]
@agileware-justin looks like it needs someone to review it to get it over the line....

@seamuslee001 once this & a couple of follow ons are merging a bunch of php8.2 test fails should fall away

**Comment by agileware-justin** [comment]
Back from hols tomorrow

**Comment by agileware-justin** [comment]
I should be able to test this today and let you know how it goes

**Comment by eileenmcnaughton** [comment]
Thanks @agileware-justin - after this is merged I'll try to do the same to the online one - which has an extra challenge or 2 (damn you secondary contribution)

**Comment by eileenmcnaughton** [comment]
@agileware-justin is this still on your list?

**Comment by agileware-justin** [comment]
Yes and no - really need the Contribution Receipt Online message tempate as that's more in line with a current project.

My testing would be based on Action Provider and Contribution Page - more interested in seeing it work with Action Provider as that's how we tend to do these things days.

**Comment by eileenmcnaughton** [comment]
@agileware-justin the changes in this PR are already made on the online contribution receipt - the membership receipts & event receipts are outstanging (including in the latter case fixing the secondary participants)

File: CRM/Contribute/WorkflowMessage/ContributionTrait.php [hunk 1]
```diff
@@ -87,6 +89,7 @@ private function getOrder(): ?CRM_Financial_BAO_Order {
    * Get bool for whether a line item breakdown be displayed.
    *
    * @return bool
+   * @noinspection PhpUnused
```
Comment (demeritcowboy) [inline CRM/Contribute/WorkflowMessage/ContributionTrait.php:92]: There are several of these added in this PR. Does it turn off "unused variable" IDE warnings? Do we want that?

File: xml/templates/message_templates/membership_offline_receipt_html.tpl [hunk 2]
```diff
@@ -100,77 +99,67 @@
...
                   </tr>
-                {/foreach}
-                {if !empty($dataArray)}
-                  {if $formValues.total_amount and $totalTaxAmount}
+
+                {if $isShowTax && '{contribution.tax_amount|raw}' !== '0.00'}
                   <tr>
                     <td {$labelStyle}>
-                      {ts}Amount Before Tax:{/ts}
+                        {ts}Amount Before Tax{/ts}:
```
[the remark this answers is not in the document]
Comment (eileenmcnaughton) [reply xml/templates/message_templates/membership_offline_receipt_html.tpl]: Ah thanks @demeritcowboy  - I fixed this string across the various templates

**Comment by agileware-justin** [comment]
Apologies for not getting time to review this one @eileenmcnaughton

File: xml/templates/message_templates/membership_online_receipt_text.tpl [hunk 3]
```diff
@@ -70,7 +70,7 @@
 {/foreach}
 
 {if !empty($dataArray)}
-{ts}Amount before Tax{/ts}: {$amount-$totalTaxAmount|crmMoney:$currency}
+{ts}Amount before Tax:{/ts} {$amount-$totalTaxAmount|crmMoney:$currency}
```
Comment (agileware-justin) [inline xml/templates/message_templates/membership_online_receipt_text.tpl:73]: Shouldn't this be ```{ts}Amount before Tax{/ts}:``` ? And not ```{ts}Amount before Tax:{/ts}```

File: xml/templates/message_templates/membership_offline_receipt_text.tpl [hunk 4]
```diff
@@ -4,73 +4,67 @@
...
-{/foreach}
+{foreach from=$lineItems item=line}
+{line.title} {$line.line_total|crmMoney|string_format:"%10s"}  {if $isShowTax && '{contribution.tax_amount|raw}' !== '0.00'} {$line.unit_price*$line.qty|crmMoney:'{contribution.currency}'|string_format:"%10s"} {if $line.tax_rate || $line.tax_amount != ""}  {$line.tax_rate|string_format:"%.2f"} %  {$line.tax_amount|crmMoney:'{contribution.currency}'|string_format:"%10s"}  {else}                  {/if}   {$line.line_total+$line.tax_amount|crmMoney|string_format:"%10s"} {/if} {$line.membership.start_date|string_format:"%20s"} {$line.membership.end_date|string_format:"%20s"}
 {/foreach}
 
-{if !empty($dataArray)}
-{ts}Amount before Tax{/ts}: {$formValues.total_amount-$totalTaxAmount|crmMoney:$currency}
+{if $isShowTax && '{contribution.tax_amount|raw}' !== '0.00'}
+{ts}Amount before Tax:{/ts} {contribution.tax_exclusive_amount}
```
Comment (agileware-justin) [inline xml/templates/message_templates/membership_offline_receipt_text.tpl:47]: Another one -> ```{ts}Amount before Tax:{/ts}```

File: xml/templates/message_templates/contribution_online_receipt_text.tpl [hunk 5]
```diff
@@ -33,7 +33,7 @@
 {/foreach}
 
 {if $isShowTax && '{contribution.tax_amount|raw}' !== '0.00'}
-{ts}Amount before Tax{/ts}: {$amount-$totalTaxAmount|crmMoney:$currency}
+{ts}Amount before Tax:{/ts} {$amount-$totalTaxAmount|crmMoney:$currency}
```
Comment (agileware-justin) [inline xml/templates/message_templates/contribution_online_receipt_text.tpl:36]: And another -> ```{ts}Amount before Tax:{/ts}```

File: xml/templates/message_templates/event_offline_receipt_text.tpl [hunk 6]
```diff
@@ -122,7 +122,7 @@
 
 {if !empty($dataArray)}
 {if $totalAmount and $totalTaxAmount}
-{ts}Amount before Tax{/ts}: {$totalAmount-$totalTaxAmount|crmMoney:$currency}
+{ts}Amount before Tax:{/ts} {$totalAmount-$totalTaxAmount|crmMoney:$currency}
```
Comment (agileware-justin) [inline xml/templates/message_templates/event_offline_receipt_text.tpl:125]: And another -> ```{ts}Amount before Tax:{/ts}```

File: xml/templates/message_templates/event_online_receipt_text.tpl [hunk 7]
```diff
@@ -142,7 +142,7 @@ You were registered by: {$payer.name}
 
 {if !empty($dataArray)}
 {if isset($totalAmount) and isset($totalTaxAmount)}
-{ts}Amount before Tax{/ts}: {$totalAmount-$totalTaxAmount|crmMoney:$currency}
+{ts}Amount before Tax:{/ts} {$totalAmount-$totalTaxAmount|crmMoney:$currency}
```
Comment (agileware-justin) [inline xml/templates/message_templates/event_online_receipt_text.tpl:145]: And another -> ```{ts}Amount before Tax:{/ts}```

File: xml/templates/message_templates/membership_offline_receipt_html.tpl [hunk 8]
```diff
@@ -100,77 +99,67 @@
...
                   </tr>
-                {/foreach}
-                {if !empty($dataArray)}
-                  {if $formValues.total_amount and $totalTaxAmount}
+
+                {if $isShowTax && '{contribution.tax_amount|raw}' !== '0.00'}
                   <tr>
                     <td {$labelStyle}>
-                      {ts}Amount Before Tax:{/ts}
+                        {ts}Amount Before Tax:{/ts}
```
[the remark this answers is not in the document]
Comment (demeritcowboy) [reply xml/templates/message_templates/membership_offline_receipt_html.tpl:142]: @agileware-justin See earlier comment: https://github.com/civicrm/civicrm-core/pull/25930#discussion_r1179193344

**Comment by demeritcowboy** [comment]
I'll give it a little run but since I don't use memberships anywhere and there's way too many variations to test this properly, I'll just focus on:
* Does it work in a basic scenario and look roughly like before.
* What happens if you have an older custom message template - does it still work.

**Comment by demeritcowboy** [comment]
If you are using an older template, and you have taxes, you get an undefined var $total_amount, and all the receipt shows for monies is "Total Tax Amount $5.00"

So possibly this should be added to the list of updated message templates that display during upgrade so people know they HAVE to update.

Oddly both with the old and new template it never shows "Amount before tax". Maybe out of scope since it seemed wrong before too. (There are also several other things that seem wrong before and after.)

Also this will need to run regen eventually.

But otherwise a basic run works.

File: xml/templates/message_templates/membership_offline_receipt_html.tpl (hunk 8 above)
[the remark this answers is not in the document]
Comment (agileware-justin) [reply xml/templates/message_templates/membership_offline_receipt_html.tpl:142]: Yep I did @demeritcowboy

[the remark this answers is not in the document]
Comment (agileware-justin) [reply xml/templates/message_templates/membership_offline_receipt_html.tpl:142]: @demeritcowboy apologies - I didn't read this document properly: "Typography is different in different languages and thus must be translated along with the string" https://docs.civicrm.org/dev/en/latest/translation/#include-typography-in-strings

Learnt something new and I suspect that all our extensions are not compliant with that rule.

So yes, my previous comments were wrong!

[the remark this answers is not in the document]
Comment (demeritcowboy) [reply xml/templates/message_templates/membership_offline_receipt_html.tpl:142]: {ts}All Good{/ts}!

[the remark this answers is not in the document]
Comment (eileenmcnaughton) [reply xml/templates/message_templates/membership_offline_receipt_html.tpl:142]: @agileware-justin yeah I thought outside was better too - until @demeritcowboy pointed it out

**Comment by eileenmcnaughton** [comment]
thanks @demeritcowboy - I'm thinking that issue with `$total_amount` would pre-date this? My preference would be to force an update on this template when the online one is also done - but if it has changed with this PR obviously it would need to be done with the merge of this PR

**Comment by eileenmcnaughton** [comment]
@demeritcowboy if I convert the membership page to use a price set I see the amount before tax

image

**Comment by demeritcowboy** [comment]
Ok with a price set it does display Amount Before Tax. By the way I'm not using a contribution page since this is the offline receipt. What I'm doing:

* Memberships - New Membership (which doesn't exist in recent versions which is confusing but you can get to the same page from an existing contact in the actions dropdown)
* Fill it out. Leave the defaults, just picking the membership type.
* Check the box for Send Receipt.
* Look at the receipt.

For the total_amount I see it only happens if you are resending the receipt by editing an existing membership and checking the box, and yes it does seem to happen both before and after.

Given how many things are a bit messed up with this to begin with, I'd say let's merge it. I'll put a PR for yet another thing wrong since it's an easy fix.

**Comment by eileenmcnaughton** [comment]
thanks @demeritcowboy I think there is some copy & paste at play - althought I would argue there should be so much copy & paste that the online & offline are the same....

**Comment by agileware-justin** [comment]
@eileenmcnaughton yes, both the offline and online message templates should be the same for contributions. Good point to raise with the product manager at the next roadmap meeting.

**Comment by eileenmcnaughton** [comment]
@agileware-justin lol - I don't think that is at the 'roadmap' level - just one of a gazillion things that would be nice to do & fall into the scratch-your-itch priority list

**Pull request merged**